File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from example_interfaces.srv import AddTwoInts
from functools import partial

class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__('add_two_ints_client')
        self.get_logger().info('Client has been started')
        self.call_add_two_ints(1, 12)
            
    def call_add_two_ints(self, a, b):
        client = self.create_client(AddTwoInts, 'add_two_ints')
        while not client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn('service not available, waiting again...')
            
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future = client.call_async(request)    
        future.add_done_callback(partial(self.call_back, a=a, b=b))        
            
    def call_back(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info('%d + %d = %d' % (a, b, response.sum))
        except Exception as e:
            self.get_logger().error('Service call failed %r' % (e,))
            
def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()
    
if __name__ == '__main__':
    main()

File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/add_two_ints.hpp>
#include <thread>

class AddTwoIntsClientNode : public rclcpp::Node
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client_cpp")
    {
        thread_ = std::thread(std::bind(&AddTwoIntsClientNode::call_add_two_ints, this, 3, 8));
    }

    void call_add_two_ints(int a, int b)
    {
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            if (!rclcpp::ok())
            {
                RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
                return;
            }
            RCLCPP_INFO(this->get_logger(), "service not available, waiting again...");
        }
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;
        auto future = client->async_send_request(request);
        auto result = future.get();
        RCLCPP_INFO(this->get_logger(), "Result of add_two_ints: %ld", result->sum);
    }
private:
std::thread thread_;
};

int main(int argc, char **argv)
{
    // Initialize the ROS system and become a node.
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    // Let the ROS run.
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}